### PR TITLE
Update explainer.md with resolved open issues

### PR DESCRIPTION
**Pull request by khushalsagar**

### CODE REVIEW

File: explainer.md
Comment (vmpstr) [inline explainer.md:53]: Presumably this would only not be satisfied (or detected as not satisfied) after style resolution

Comment (vmpstr) [reply explainer.md:53]: It's also unclear to me if we're requiring contain: paint or simply a property which adds paint containment. I'll file a separate issue for this.

File: explainer.md
Comment (vmpstr) [inline explainer.md:92]:
```suggestion
In the example transition, the content of the header cross-fades from Page-A to Page-B. An even smoother transition could be achieved by also animating the site title and avatar 'chip' independently. When a developer offers nested shared elements, the state can be captured in one of the following two modes:
```

File: explainer.md
Comment (vmpstr) [inline explainer.md:94]:
```suggestion
#### Flat List Mode
```

File: explainer.md
Comment (vmpstr) [inline explainer.md:99]:
```suggestion
#### Hierarchy Mode
```

File: explainer.md
Comment (vmpstr) [inline explainer.md:100]:
```suggestion
In this mode, the hierarchy of shared elements is preserved along with a transform which is relative to its nearest ancestor shared element instead of the viewport. When used with the "computed style + content image" mode, this also preserves other hierarchical properties (clip, inherited effects) that a shared element inherits from its ancestor. 
```

File: explainer.md
Comment (vmpstr) [inline explainer.md:102]: Is this a separate mode solely for minimizing the scope of the v1?

File: explainer.md
Comment (ianvollick) [inline explainer.md:238]: nit: is this an unintentional edit?

File: explainer.md
Comment (vmpstr) [inline explainer.md:128]: not sure what "addressable" means in this context, but this was there before

File: explainer.md
Comment (vmpstr) [inline explainer.md:156]: I don't really like the name uber-root for official explainers. I don't have a better name, but maybe we can just not name it?

File: explainer.md
Comment (vmpstr) [inline explainer.md:160]: Perhaps root stacking context can be "page root stacking context" or something that makes it clear that it's a root for all things that the page itself may paint. Although technically the transition stacking context's elements are also styled by the page, so nevermind...

Comment (khushalsagar) [reply explainer.md:160]: It's actually better to call it "root stacking context" to make it obvious that we're talking about the existing concept of a page's root stacking context?

File: explainer.md
Comment (khushalsagar) [reply explainer.md:156]: I mostly did it for the diagram below to make the proposed hierarchy obvious. In general, we have a lot of naming in this explainer that needs to change. :)

File: explainer.md
Comment (khushalsagar) [reply explainer.md]: How about now? The aim is to say that developers can access them as pseudo elements.

File: explainer.md
Comment (khushalsagar) [reply explainer.md:102]: That's the main reason, yes. Other reason is that we can choose to not support it for the out of process cross-origin implementation. But the primary motivation is to scope it out of v1.

File: explainer.md
Comment (khushalsagar) [reply explainer.md:53]: Yes, we can detect it during style resolution and abort in the same frame.

File: explainer.md
Comment (khushalsagar) [reply explainer.md]: Yes, my bad. Fixed.

File: explainer.md
Comment (vmpstr) [reply explainer.md:53]: Do you mind saying this? "At any point during transition" sounds too vague

Comment (khushalsagar) [reply explainer.md]: Done.

File: explainer.md
Comment (ianvollick) [reply explainer.md:102]: Would it be possible to support this cross-origin if hierarchical info were sent to wherever it is that interpolation happens?

Comment (khushalsagar) [reply explainer.md:102]: Yeah, it's definitely possible. We'll need to build a data structure that preserves the transform, clip and stacking context hierarchy of these elements. All of those should be the same since with paint containment each shared element creates a stacking context anyway.

One aspect worth pointing out is that this hierarchy mode will only retain positioning and clipping from an ancestor. If the ancestor has effects like opacity (which should continue to be inherited by the descendants) then we have to ensure that the opacity is "live" and not baked into the snapshot. So it applies to the shared element's image and also to the images of its descendants.

@chrishtr has also mentioned not painting visual effects into the snapshot. This is a divergence from the element() function behaviour but maybe it makes sense to have the "As a single image" mode be "As style for visual effects + single image" mode instead. The styles for visual effects which apply to descendants are copied to the container.

We can support "As style for visual effects + single image" + hierarchy for cross-origin but supporting "As the element's computed style + content image" for cross-origin will be too difficult.

**Review by vmpstr — APPROVED** [review]

File: explainer.md
Comment (khushalsagar) [reply explainer.md:102]: I've filed a WICG issue for the question about visual effects above.

Comment (ianvollick) [reply explainer.md:102]: Thanks for the extra info. I was just hoping to clarify the point raised earlier and didn't mean to hold up the review. I will stamp and defer to Vlad's review.

**Review by ianvollick — APPROVED** [review]

File: explainer.md
Comment (jakearchibald) [inline explainer.md:59]: This change feels like an error. It means:

- Transition elements will appear at 0,0 all stacked on top of each other, since viewport position is no longer captured.
- Transition elements will have to be sized manually, since their dimensions are no longer captured.
- If the transition element is rotated 45 degs, that will be baked into the capture.

File: explainer.md
Comment (jakearchibald) [inline explainer.md:95]: I can't get my head around this. Is the word "content" being used correctly here?

```html
<div>Hello</div>
```

The content of that `div` is the `Hello` text node. If the div has a border or background colour, that isn't part of the content.

File: explainer.md
Comment (jakearchibald) [inline explainer.md:94]: I'm not sure what this section is describing. You can try to explain it to me in a chat 😄

File: explainer.md
Comment (jakearchibald) [reply explainer.md:160]: It isn't clear to me what this is vs "root image"

File: explainer.md
Comment (jakearchibald) [inline explainer.md:302]: This API doesn't exist. Is it being proposed here?

File: explainer.md
Comment (jakearchibald) [inline explainer.md:308]: This feels like it's saying shadow DOM doesn't already exist in the platform.

File: explainer.md
Comment (jakearchibald) [reply explainer.md:94]: (we chatted and I think I understand it now. I'll add to this PR)

File: explainer.md
Comment (jakearchibald) [reply explainer.md:95]: (we chatted and I think I understand it now. I'll add to this PR)

File: explainer.md
Comment (jakearchibald) [reply explainer.md:160]: (we chatted and I think I understand it now. I'll add to this PR)

File: explainer.md
Comment (khushalsagar) [reply explainer.md:302]: You mean the .animate API on CSSPseudoElement? Yeah, I was proposing that.

I've noticed a bunch of APIs for pseudo elements work by adding it as an argument, like window.getComputedStyle(element, "::after"). This is likely there wasn't an object in script for a pseudo element when these were added. Now that we have one, why not add these APIs to that interface.

File: explainer.md
Comment (khushalsagar) [reply explainer.md:308]: The concept that's missing is creating a shadow DOM which has no element to be its shadow host.

File: explainer.md
Comment (hvanops) [reply explainer.md:102]: "This is proposed as a separate mode to minimize the scope for 'v1' of the feature." This makes it sound like separating into two modes minimizes the scope. Can we clarify that only Flat List mode will be targeted for v1 to minimize scope.

Comment (khushalsagar) [reply explainer.md:102]: @jakearchibald mentioned that he'll reword this section to make the details for the 2 modes clearer. Will make sure that flat list being the only mode for v1 is obvious after his edits.

File: explainer.md
Comment (jakearchibald) [inline explainer.md:96]: @khushalsagar I reworked this idea a bit so it feels less like a v2 switch, and more like opting into a new positioning space like `position: relative`. Let me know what you think. I went back and forth on elements of this, so I'm absolutely open to tweaks and better ideas.

File: explainer.md
Comment (jakearchibald) [inline explainer.md:114]: @khushalsagar in my last PR I had the offered elements absolutely positioned in a fixed position container, and we switched it back to everything being fixed position. However, I think we need absolute positioning else the nesting doesn't make sense. We want transition elements to be positioned relative to 0,0 of their transition container, which may be the viewport, but may not be. Making them fixed position would put everything at 0,0 in viewport space, which doesn't seem to work.

File: explainer.md
Comment (jakearchibald) [inline explainer.md:125]: I've added a new wrapper 😬, but it seems necessary to handle nesting. I guess we could omit this wrapper for transition elements that aren't a 'transition container', but is it better to always have it for consistency?

File: explainer.md
Comment (ianvollick) [reply explainer.md:114]: Regarding fixed positioning; the containing block for a fixed position element is usually the viewport, but not always (eg, setting `transform:` on an element establishes a containing block for all positioned descendants, including fixed [1]). If we want to describe elements as being fixed with respect to its nearest shared ancestor, I think we could use the same type of prose.

[1] https://www.w3.org/TR/css-transforms-1/#:~:text=the%20transform%20property%20also%20causes%20the%20element%20to%20establish%20a%20containing%20block%20for%20all%20descendants.

File: explainer.md
Comment (khushalsagar) [inline explainer.md:94]:
```suggestion
When an element is captured, its painting is 'lifted' from the parent offered element, which may be the root. It's analogous to the nested elements having opacity:0 in the parent element's image. This is similar to how browsers handle composited elements.
```

File: explainer.md
Comment (khushalsagar) [reply explainer.md:96]: The idea sounds interesting. Do you have any example in mind that could make a case for your parent being different from the nearest ancestor shared element? And also for having a completely flat list.

Because the alternate choice we were debating was to have every offered element be a 'transition container'. With the assumption that there is no use-case where the developer would want to change this hierarchy or have a flat list. If we could point to a use-case then it would help with making an argument for the flat mode being a useful mode and not just a stop-gap until we have hierarchy.

File: explainer.md
Comment (khushalsagar) [inline explainer.md:104]:
```suggestion
Alternatively, the developer can make an offered element a 'transition container'. Offered elements will be nested within their closest transition container, and the cached transform will position the element within that container rather than the viewport. This is similar in spirit to how `position: relative` becomes the containing block for absolutely positioned elements.
```

File: explainer.md
Comment (khushalsagar) [inline explainer.md:105]: We should also mention the clipping and effects being preserved if the developer uses style + image mode when capturing the element. So for example if your ancestor shared element has opacity: 0.5, it won't be preserved if you have a flat list. But with this mode, since the hierarchy is preserved the visual effects continue to be inherited.

File: explainer.md
Comment (khushalsagar) [inline explainer.md:114]:
```suggestion
The captured elements are displayed using a tree of elements. The root of this tree is a viewport-filling container with fixed position. The offered elements from Page-A are positioned absolutely at (0,0), nested according to their closest 'transition container' or the root if there's no parent container, and moved into their previous viewport-relative (or transition container relative) positions using the cached transform. Their content is painted on top of Page-B which ensures that the user continues to see Page-A's visuals as Page-B is loading. Note that this may not reproduce the exact rendering on Page-A. For example, the relative paint order of shared elements is preserved in this tree rendered on top of Page-B. But if a shared element was occluded by another element, the latter is painted into the root's image unless it is also offered as a shared element.
```

Comment (khushalsagar) [reply explainer.md:114]: Using a fixed position root for everything SGTM. We actually also needed that to describe the paint order of transition elements (using z-index). The transition DOM I'm expecting looks something like this :

transition root (position: fixed)
|_ container (position: absolute)
    |_ wrapper
        |_ old image
        |_ new image

Where the transition root has a flat list of containers for each shared element + the root image. If one of the shared elements is offered as the "transition container" for nesting then you'd have something like this :

transition root (position: fixed)
|_ container (position: absolute)
    |_ wrapper
    |   |_ old image
    |   |_ new image
    |_ container (position: absolute)
        |_ wrapper
            |_ old image
            |_ new image

So the container has nested container elements inside it. Does that sound right to you?

File: explainer.md
Comment (khushalsagar) [reply explainer.md:125]: My comment above touches on this DOM structure. Let's discuss it there? I don't think we should need to a new wrapper type but I might be missing the case you had in mind.

File: explainer.md
Comment (ianvollick) [reply explainer.md:114]: Say we have a transition element that's position:fixed -- will its containing block still be the viewport? If that's true, it seems like we could be in a situation where a transition element's transform/positional parent is different than its effect parent. I.e., we'd need something like property trees to represent the relationships and things become far more complex. Could we instead make it be the case that all non-leaf nodes in this transition hierarchy establish a containing block for all positioned descendants?

Also, since this seems a bit involved, could we perhaps add a TODO for clarifying the details of the hierarchical mode and discuss this on a separate, dedicated PR?

File: explainer.md
Comment (khushalsagar) [inline explainer.md:177]:
```suggestion
This allows using the output of the root stacking context to provide a live root image for Page-B. The transition stacking context maps to the transition-root element. An alternate approach to this was to paint the transition-root in the [top layer](https://fullscreen.spec.whatwg.org/#top-layer). But that made it difficult to support transitions when there is other content in the top layer (fullscreen elements, dialog) and to ensure effects on the root element which are applied to the root stacking context (background-color, filter) are captured in the root image. See [issue](https://github.com/WICG/shared-element-transitions/issues/74) for detailed discussion.
```

Can clarify what the transition stacking context is if we specify the transition-root element in the DOM structure.

File: explainer.md
Comment (khushalsagar) [reply explainer.md:114]: "we could be in a situation where a transition element's transform/positional parent is different than its effect parent". Do you mean if the developer overrides anything in this DOM to be position:fixed? If we're building this using DOM elements (or pseudo-elements) then why would it be difficult if the positional parent is different than the effect or clip parent. Are you concerned about the "out of process" implementation getting complicated?

Comment (ianvollick) [reply explainer.md:114]: > Do you mean if the developer overrides anything in this DOM to be position:fixed?

Yep, that's right.

> If we're building this using DOM elements (or pseudo-elements) then why would it be difficult if the positional parent is different than the effect or clip parent. Are you concerned about the "out of process" implementation getting complicated?

I am, yeah. If engines want to support this outside of renderer processes, than this complexity would need to be replicated.

Comment (khushalsagar) [reply explainer.md:114]: The default hierarchy and styles set up by the UA (as described above) should ensure an element's transform parent is the same as its effect/clip parent. This only becomes an issue if the developer can modify the style on these elements and that's allowed for the renderer process implementation (same-origin) only. So it won't impact the out of process implementation.

Comment (ianvollick) [reply explainer.md:114]: Could this not be impacted by the styles on transition elements even in the implicit set up for cross-origin? Or are you saying that we'd override the positioning to be `position:absolute` in that case and the only way it would change is if the developer explicitly changed that via JS on the destination page?

Comment (khushalsagar) [reply explainer.md:114]: Yes. We're setting the position for all container elements to be position: absolute with a UA stylesheet. Developers can change it only for same-origin via CSS or script.

Comment (ianvollick) [reply explainer.md:114]: Got it. Thanks!

File: explainer.md
Comment (jakearchibald) [reply explainer.md:177]: fwiw, I didn't edit this section. I'm not sure I fully understand it.

File: explainer.md
Comment (jakearchibald) [reply explainer.md]: Seems ok, but I'm a little worried we're using 'root' a lot to mean different things.

File: explainer.md
Comment (jakearchibald) [reply explainer.md:96]: I tried to include a bit of this in the explainer. A flat list allows you to scale the parent without stretching the children, or needing to provide a counter-transform to provide that. A flat list also makes it easier if a child element moves containers between pages.

File: explainer.md
Comment (jakearchibald) [reply explainer.md]: @khushalsagar lol yeah the DOM model I had doesn't make sense. Not sure what I was thinking.

Comment (jakearchibald) [reply explainer.md]: I reworked the DOM stuff (which meant mostly reverting it back to what it was.

**Review by jakearchibald — APPROVED** [review]

**Pull request merged**
